src/prime_rl/trainer/rl/data.py

```diff
@@ -39,18 +39,70 @@ def micro_batch_to_tensor(micro_batch: MicroBatch) -> TensorMicroBatch:
 
 
 class FakeDataLoader:
-    def __init__(self, config: FakeDataLoaderConfig, seq_len: int):
+    def __init__(self, config: FakeDataLoaderConfig, seq_len: int, dp_world_size: int):
+        self.world = get_world()
+        self.dp_world_size = dp_world_size
+        self.non_dp_world_size = self.world.world_size // self.dp_world_size
+        self.dp_rank = self.world.rank // self.non_dp_world_size
+
         self.batch_size = config.batch_size
-        self.num_micro_batches = self.batch_size // get_world().world_size
+        self.num_micro_batches = self.batch_size // self.dp_world_size
         self.seq_len = seq_len
+        self.generate_samples = config.generate_samples
+        self.batch_counter = 0
 
     def wait_for_batch(self) -> None:
         return
 
     def get_batch(self) -> list[TensorMicroBatch]:
-        return [self._get_micro_batch() for _ in range(self.num_micro_batches)]
+        if not self.generate_samples:
+            get_micro_batch_fn = self._get_micro_batch
+        else:
+            get_micro_batch_fn = self._get_document_micro_batch
+
+        # This is a pretty ugly hack to ensure that all CP ranks in a data parallel group receive the same micro batch.
+        micro_batches = []
+        for micro_batch_idx in range(self.num_micro_batches):
+            seed = self.dp_rank * 1000000 + self.batch_counter * 1000 + micro_batch_idx
+            generator = torch.Generator().manual_seed(seed)
+            micro_batches.append(get_micro_batch_fn(generator))
```

Comment:
seems very ugly indeed haha

Reply:
oh okay but then why do we even need it here ? for fake data we don't really care they have the same data right ?